openrewrite/src/javascript/parserUtils.ts

```diff
@@ -180,3 +180,51 @@ export function binarySearch<T>(arr: T[], target: T, compare: (a: T, b: T) => nu
     }
     return ~low;  // Element not found, return bitwise complement of the insertion point
 }
+
+export function hasFlowAnnotation(sourceFile: ts.SourceFile) {
+    if (sourceFile.fileName.endsWith('.js') || sourceFile.fileName.endsWith('.jsx')) {
+        const comments = sourceFile.getFullText().match(/\/\*[\s\S]*?\*\/|\/\/.*(?=[\r\n])/g);
+        if (comments) {
+            return comments.some(comment => comment.includes("@flow"));
+        }
+    }
+    return false;
+}
+
+export function checkSyntaxErrors(program: ts.Program, sourceFile: ts.SourceFile) {
+    const diagnostics = ts.getPreEmitDiagnostics(program, sourceFile);
+    // checking Parsing and Syntax Errors
+    let syntaxErrors : [errorMsg: string, errorCode: number][] = [];
+    if (diagnostics.length > 0) {
+        const errors = diagnostics.filter(d =>  (d.category === ts.DiagnosticCategory.Error) && isCriticalDiagnostic(d.code));
+        if (errors.length > 0) {
+            syntaxErrors = errors.map(e => {
+                let errorMsg;
+                if (e.file) {
+                    let {line, character} = ts.getLineAndCharacterOfPosition(e.file, e.start!);
+                    let message = ts.flattenDiagnosticMessageText(e.messageText, "\n");
+                    errorMsg = `${e.file.fileName} (${line + 1},${character + 1}): ${message}`;
+                } else {
+                    errorMsg = ts.flattenDiagnosticMessageText(e.messageText, "\n");
+                }
+                return [errorMsg, e.code];
+            });
+        }
+    }
+    return syntaxErrors;
+}
+
+const additionalCriticalCodes = new Set([
+    // Syntax errors
+    17019, // "'{0}' at the end of a type is not valid TypeScript syntax. Did you mean to write '{1}'?"
+    17020, // "'{0}' at the start of a type is not valid TypeScript syntax. Did you mean to write '{1}'?"
+
+    // Other critical errors
+]);
+
+const excludedCodes = new Set([1039, 1064, 1101, 1107, 1111, 1155, 1166, 1170, 1183, 1203, 1207, 1215, 1238, 1239, 1240, 1241, 1244, 1250,
```

Comment:
Are there constants with readable names for all these codes?

Reply:
I can put here the code description.
The full list is available here https://github.com/microsoft/TypeScript/blob/main/src/compiler/diagnosticMessages.json